streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimestampedKeyValueStore.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.StateSerdes;
+import org.apache.kafka.streams.state.TimestampedKeyValueStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+public class InMemoryTimestampedKeyValueStore<K, V> implements TimestampedKeyValueStore<K, V> {
```

Comment:
Just curious, why are we implementing this as a separate class instead of wrapping the existing InMemoryKeyValueStore?

Reply:
In the original KIP, we added new method to interface `TimestampedKeyValueStore`, eg:
```
public synchronized ValueAndTimestamp<V> putIfAbsent(final K key,
                                                     final V value,
                                                     final long timestamp)
```
Hence, this PR is a little out dated, and I agree that we don't need `InMemoryTimestampedKeyValueStore` any longer -- I'll cleanup and rebase the PR accordingly.